### DISCUSSION TITLE
Default to closed structs for {} syntax?

### DISCUSSION THREAD

**Discussion by Kinrany** [cue-lang/cue#1149] [Q&A]
Couldn't find the reasoning behind making lists closed but structs open by default.

Some reasons why one might want structs to be closed by default:

* Consistent syntax between lists and structs: use `...` for both instead of a separate `close()`
* Natural separation between `a: b: c`, which describes one field of `a`, and `a: { b: c }`, which describes the whole `a`
* When closedness matters, it's more common to prefer closed structs since it provides more information to the consumers
* When closedness doesn't matter, it might affect performance since a closed struct needs a fixed set of fields

I'm not really using CUE for anything yet, just sharing something that seemed surprising :)

**Comment by myitcv** [comment]
@Kinrany welcome to the CUE project! On the basis this isn't a specific bug/feature request I've converted your issue to a discussion.

In any case, I'll defer to @mpvl for thoughts on this.

**Comment by verdverm** [comment]
A few thoughts that come to mind

- closedness is one of the trickier issues in the evaluator, some really interesting code to go along with it!
- definitions are the (recursively) closed by default structs, would closed by default structs be semantically different? (based on the example it seems so, but then there are semantic differences between what are today semantic equivalences (2nd bullet))
- I've definitely become accustomed to the open by default. Any existing code would break so this is no small change. Existing code, or anything that wants to be recursively open, would become littered with `...`, i.e. require ellipses at every level. It would be hard to notice if one was missing or extra. It requires looking at the end of a scope, rather than the beginning, which compounds the strain on readability.

```
ex: {
  a: {
    aa: {
      aaa: _
      aab: _
      ...
    }
    ab: {
      aba: _
      abb: _
      ...
    }
  }
  b: {
    ba: {
      baa: _
      bab: _
      ...
    }
    bb: {
      bba: _
      bbb: _
      ...
    }
    ...
  }
  ...
}
```

**Reply by Kinrany** [reply]
Thanks for the thoughts!

> would closed by default structs be semantically different?

Indeed those are the semantics I want, except for the difference between `a: b: "x"` and `a: { b: "x" }`. Perhaps this would also allow dissolving definitions into a combination of closed `{}` structs and private `#a` fields? Having less concepts seems like a win.

> Any existing code would break so this is no small change.

Yep! Should be relatively easy to do automatically in most cases since this only affects `{}`: need to replace `a: { b: }` with `a: { ..., b: "x" }` or `a: b: "x"` and `close({ a: "x" })` with `{ a: "x" }`. And easier to do now while breaking changes are still OK, if this turns out to be a good idea.

> Existing code, or anything that wants to be recursively open, would become littered with ..., i.e. require ellipses at every level.

This seems symmetric and a question of which semantics are needed more often (ignoring the cases where it doesn't matter). It should be possible to separately have `deep_open()` if closed semantics are preferred in general, but nested open structs are also common.

With `a: b: c: "x"` remaining open it's possible to write the example above as the following, and I'm not sure it's worse:

```cue
ex: a: aa: aaa: _
ex: a: aa: aab: _
ex: a: ab: aba: _
ex: a: ab: abb: _
ex: b: ba: baa: _
ex: b: ba: bab: _
ex: b: bb: bba: _
ex: b: bb: bbb: _
```

> It requires looking at the end of a scope, rather than the beginning, which compounds the strain on readability.

Adding `...` in the beginning works for definitions, should work for closed structs too

**Reply by verdverm** [reply]
Note that in my example, you missed one of the fields which is closed, so the line-by-line rewrite is not equivalent. (case in point?)

I wonder if this would complicate the closedness checking code, which relies on knowing if something is recursively closed for some optimizations. There are certainly a lot of other things that need to change, tools like trim and flow. I wonder how this would affect comprehension as well...?

in something like `a: { b: c: _ }` is only `a` closed, or are `b` and `c` closed as well? If also closed, then is there any difference with a definition?

Given we have a closed by default feature available, it's nice to have an open by default rather than something that is neither.

**Reply by Kinrany** [reply]
> Note that in my example, you missed one of the fields which is closed, so the line-by-line rewrite is not equivalent. (case in point?)

Ah. Yes, I agree that having lots of ellipses isn't great for deep nesting.

> in something like a: { b: c: _ } is only a closed, or are b and c closed as well? If also closed, then is there any difference with a definition?

Just `a`! Would have to be `a: { b: { c: _ } }` for both. (Which is kinda ugly, but the alternative is `a: close({ b: close({ c: _ }) })`).

> Given we have a closed by default feature available, it's nice to have an open by default rather than something that is neither.

Aren't definitions supposed to be used as schemas? It seems like this should be doable with the basics given types and values are explicitly the same thing in CUE.

**Comment by mpvl** [comment]
It's mostly for pragmatic reasons. In practice, schema will be defined separately from data. Data are almost always partial configurations. It would be awkward to add `...` everywhere. In general, data writers should not really be concerning themselves with closedness.

So closedness was added only to definitions only for this reason. Note that closedness is really a "soft" constraint as well that only serves the purpose of detecting typos. In general, for APIs (schema) to allow for growth and extendibility, they should be defined considered to be open.

That said, you do bring up a good point. One could argue that lists should be _open_ by default outside of definitions. That would actually make a lot of sense. The current semantics predates the existence of definitions. But it actually seems sensible to treat lists like that.
